src/error.rs

```diff
@@ -1,84 +1,515 @@
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum UnsupportedErrorKind {
+    /// The required color type can not be handled.
+    Color(ExtendedColorType),
+    /// An image format is not supported.
+    Format(ImageFormatHint),
+    /// Some feature specified by string.
+    /// This is discouraged and is likely to get deprecated (but not removed).
+    GenericFeature(String),
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+#[derive(Debug)]
+pub struct EncodingError {
+    format: ImageFormatHint,
+    message: String,
+    underlying: Option<Box<dyn Error + Send + Sync>>,
+}
 
-    /// There is not enough memory to complete the given operation
+#[derive(Debug)]
+pub struct ParameterError {
+    kind: ParameterErrorKind,
+    underlying: Option<Box<dyn Error + Send + Sync>>,
+}
+
+/// Details how a parameter is malformed.
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum ParameterErrorKind {
+    /// Repeated an operation for which error that could not be cloned was emitted already.
+    FailedAlready,
+    /// The dimensions passed are wrong.
+    DimensionMismatch,
+    /// A string describing the parameter.
+    /// This is discouraged and is likely to get deprecated (but not removed).
+    Generic(String),
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+#[derive(Debug)]
+pub struct DecodingError {
+    format: ImageFormatHint,
+    message: Option<String>,
+    underlying: Option<Box<dyn Error + Send + Sync>>,
+}
+
+#[derive(Debug)]
+pub struct LimitError {
+    kind: LimitErrorKind,
+    // do we need an underlying error?
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[allow(missing_copy_implementations)] // Might be non-Copy in the future.
+pub enum LimitErrorKind {
+    DimensionError,
     InsufficientMemory,
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+/// A best effort representation for image formats.
+#[derive(Clone, Debug, Hash, PartialEq)]
+pub enum ImageFormatHint {
+    /// The format is known exactly.
+    Exact(ImageFormat),
+
+    /// The format can be identified by a name.
+    Name(String),
+
+    /// A common path extension for the format is known.
+    PathExtension(std::path::PathBuf),
+
+    /// The format is not known or could not be determined.
+    Unknown,
+
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+#[allow(non_upper_case_globals)]
+#[allow(non_snake_case)]
+impl ImageError {
+    pub(crate) const InsufficientMemory: Self =
+        ImageError::Limits(LimitError {
+            kind: LimitErrorKind::InsufficientMemory,
+        });
+
+    pub(crate) const DimensionError: Self =
+        ImageError::Parameter(ParameterError {
+            kind: ParameterErrorKind::DimensionMismatch,
+            underlying: None,
+        });
+
+    pub(crate) const ImageEnd: Self =
+        ImageError::Parameter(ParameterError {
+            kind: ParameterErrorKind::FailedAlready,
+            underlying: None,
+        });
+
+    pub(crate) fn UnsupportedError(message: String) -> Self {
```

Comment:
Very nice trick! This should allow to incrementally improve errors in the rest of the library, however we must ensure that these functions are not misused.

Reply:
Should they be private as here, or part of the public interface and deprecated, and if so, maybe with a changed name? It's hard to say if it not offering them makes it more enticing to properly switch, or to copy the internal implementation anyways but be mad about it not being available..

Reply:
In 7df7d73b2c3c1dba7621b9bd1dc99dd1bbe39898, where I added the public constructors for the error representations, I chose the names to encourage providing an underlying error with the short name `new` and gave a longer name to all other constructors. This basically resolves this question as keeping these functions private.